lib/dice/src/lib.rs

```diff
@@ -0,0 +1,225 @@
+
+fn get_cdi_reg_slice() -> &'static mut [u32] {
+    // SAFETY: Dereferencing this raw pointer is necessary to read the CDI
+    // from the LPC55 DICE registers. This pointer will always reference a
+    // valid memory region as this is where the LPC55 maps the DICE
+    // registers. Support from the lpc55-pac would move the unsafe code
+    // into the pac library:
+    // https://github.com/lpc55/lpc55-pac/issues/15
+    // https://github.com/lpc55/lpc55-pac/pull/16
+    unsafe {
+        slice::from_raw_parts_mut(
+            REG_ADDR_NONSEC as *mut u32,
+            SEED_LENGTH / mem::size_of::<u32>(),
+        )
+    }
+}
+
+/// NXP LPC55 UM 11126 §4.5.74 states: "Once CDI is computed and consumed,
+/// contents of those registers will be erased by ROM." Testing has shown
+/// however that the DICE registers are not cleared after they're read.
+/// This type is for accessing and clearing the NXP DICE registers (aka the
+/// CDI). To ensure the DICE registers are cleared, this type derives the
+/// ZeroizeOnDrop trait. When an instance of this object goes out of scope
+/// the register is cleared through the slice held.
+#[derive(Zeroize, ZeroizeOnDrop)]
+struct CdiReg(&'static mut [u32]);
+
+impl Default for CdiReg {
+    fn default() -> Self {
+        Self(get_cdi_reg_slice())
+    }
+}
+
+impl CdiReg {
+    fn is_clear(&self) -> bool {
+        if self.0.iter().all(|&w| w == 0) {
+            true
+        } else {
+            false
+        }
+    }
+}
+
+pub trait SeedBuf {
+    fn as_bytes(&self) -> &[u8; SEED_LENGTH];
+}
+
+/// This type is a thin wrapper around a byte array holding the CDI.
+/// It's populated from the DICE registers by calling the 'from_reg'
+/// constructor. The DICE registers are cleared immediately after the
+/// CDI is read out from them and future attempts to construct a Cdi
+/// instance through 'from_reg' will return None.
+#[derive(Zeroize, ZeroizeOnDrop)]
+pub struct Cdi([u8; SEED_LENGTH]);
+
+impl SeedBuf for Cdi {
+    fn as_bytes(&self) -> &[u8; SEED_LENGTH] {
+        &self.0
+    }
+}
+
+impl Cdi {
+    /// If DICE is disabled this function will return None. Otherwise it
+    /// copies the contents of the DICE registers into a Cdi instance that's
+    /// returned to the caller before clearing the DICE registers. This side
+    /// effect causes subsequent calls to this function to return None.
+    pub fn from_reg() -> Option<Self> {
+        let mut cdi = [0u8; SEED_LENGTH];
+        // If the CDI register hasn't already been cleared it will be when
+        // this instance goes out of scope.
+        let cdi_reg = CdiReg::default();
+
+        // When registers holding CDI have been cleared / zeroed return None
+        // to prevent unsuspecting consumers from deriving keys from 0's.
+        if cdi_reg.is_clear() {
+            return None;
+        }
+
+        for (dst, src) in cdi
+            .chunks_exact_mut(mem::size_of::<u32>())
+            .zip(cdi_reg.0.as_ref())
```

Comment:
impl SeedBuf on CdiReg instead of cdi_reg.0.as_ref()?

Reply:
CdiReg is a slice of u32s not bytes like the SeedBufs. The type exists pretty much exclusively to ensure the registers that it wraps are zeroed when it's dropped.